patterns/small-crates.md

```diff
@@ -0,0 +1,31 @@
+# Prefer small crates
+
+## Description
+
+Prefer small crates that do one thing well.
+
+Cargo and crates.io make it easy to add third-party libraries, much more so than in say C or C++. We should take advantage of this tooling, and use smaller, more fine-grained dependencies.
+
+## Advantages
+
+* Small crates are easier to understand and audit.
+* Since the compilation unit of Rust is the crate, splitting a project into multiple crates can allow more of the code to be built in parallel.
+* If published on crates.io, a crate can be reused by other projects.
+
+## Disadvantages
+
+* Too many crates can lead to code that is hard to follow.
+* This can lead to "dependency hell", where a project depends on multiple conflicting versions of a crate at the same time.
```

Comment:
I've seen the issue with Piston, e.g. PistonDevelopers/Piston-Tutorials#76. If two crates depend on different versions of the same trait, then we might get "trait not implemented" errors when using them together.

That kind of problem is pretty rare though, and it's easy to mitigate (re-export dependent crates). I wonder if "re-export dependencies" should be another pattern.